orchard/modules/_branch.py

```diff
@@ -6,6 +6,228 @@
 # ----------------------------------------------------------------------------
 
+import pprint
+import yaml
+from pathlib import Path
+import os
 
-class Branch:
-    pass
+
+# Gets a named module from a link or config file
+def get_module(file_data, name):
+    return next((item for item in file_data['modules'] if item['name'] == name), None)
+
+
+# Returns the depth level of a given module based on the link file
+def depth_level(link_file_data, module):
+    current_depth = 0
+    if 'dependencies' in module:
+        for mod_name in module['dependencies']:
+            current_depth = max(current_depth, depth_level(link_file_data, get_module(link_file_data, mod_name)))
+    return current_depth + 1
+
+
+# Returns true if an argument causes branching in a link module
+def is_branching_arg(link_file_mod, argname):
+    for arg in link_file_mod['arguments']:
+        if arg['name'] == argname:
+            if 'isBranch' in arg and not arg['isBranch']:
+                return False
+            else:
+                return True
+    return True
+
+
+# Goes through a list and returns true if the given argument branches, recursively checking in 'exclusive' blocks
+def is_branching_in_list(arglist, name):
+    for arg in arglist:
+        if 'exclusive' in arg:
+            if not is_branching_in_list(arg['exclusive'], name):
+                return False
+        if 'name' not in arg:
+            continue
+        if arg['name'] == name:
+            if 'isBranch' in arg and not arg['isBranch']:
+                return False
+            else:
+                return True
+    return True
+
+
+def check_args(module_1, module_2, link_module):
+    # For each argument in the first module
+    if 'arguments' in module_1:
+        if 'arguments' not in module_2:
+            return False
+        for arg in module_1['arguments']:
+            argname = list(arg.keys())[0]
+            if not is_branching_arg(link_module, argname):
+                continue
+            # Get the matching dictionary entry in the second config structure
+            match = next((item for item in module_2['arguments'] if list(item.keys())[0] == argname), None)
+            # If we didn't find a match then they obviously don't match
+            if match is None:
+                return False
+            if arg[argname] != match[argname]:
+                return False
+    return True
+
+
+def check_optionals(module_1, module_2, link_module):
+    if 'optionals' in module_1:
+        if 'optionals' not in module_2:
+            return False
+        # Check the normal optionals
+        for arg in module_1['optionals']:
+            argname = list(arg.keys())[0]
+            if not is_branching_in_list(link_module['optionals'], argname):
+                continue
+            # Get the matching dictionary entry in the second config structure
+            match = next((item for item in module_2['optionals'] if list(item.keys())[0] == argname), None)
```

Comment:
Same as above, RE: dictionary ordering. These might be able to be turned into one function, just noticed how similar they are.